apps/workshop/projects/Adax/modules/AdaxLiquidity.ts

```diff
@@ -0,0 +1,123 @@
+import type { ModuleDefinitionInterface, TokenDetail, TokenExtra } from '@defiyield/sandbox';
+import { CARDANO_COIN_ADDRESS } from '../../../common/constants/cardano.constants';
+import { ADAX_POLICY, AVAILABLE_POOLS } from '../helpers/constants';
+import { accountAssets } from '../helpers/queries';
+
+export const AdaxLiquidity: ModuleDefinitionInterface = {
+  name: 'AdaxLiquidity',
+  chain: 'cardano',
+  type: 'pools',
+
+  async preloadTokens() {
+    const contacts = Array.from(AVAILABLE_POOLS.keys());
+    const assets = Array.from(AVAILABLE_POOLS.values());
+
+    return contacts.concat(...assets);
+  },
+
+  async fetchMissingTokenDetails({ address }) {
+    const underlying = AVAILABLE_POOLS.get(address);
+    if (Array.isArray(underlying)) {
+      return <TokenDetail>{
+        decimals: 0,
+        address,
+        underlying,
+      };
+    }
+    return void 0;
+  },
+
+  async fetchMissingTokenPrices({ assets, axios, BigNumber, allAssets }) {
+    const {
+      data: { pools },
+    } = await axios.post('https://amm-api.adax.pro', { endpoint: 'getPools' });
+
+    const adaToken = allAssets.find((token) => token.address === CARDANO_COIN_ADDRESS);
+    const poolMap = new Map(
+      pools.map((pool: any) => [`${ADAX_POLICY}.${pool.pool_nft_name}`, pool]),
+    );
+
+    return assets
+      .filter((token) => token.address.startsWith(ADAX_POLICY))
+      .map((token) => {
+        const pool: any = poolMap.get(token.address);
+        const secondTokenAddress = `${pool.asset_b.symbol}.${pool.asset_b.name}`;
+        const secondToken = allAssets.find((token) => token.address === secondTokenAddress);
+        let tvl = new BigNumber(0);
+        let price = new BigNumber(0);
+
+        if (adaToken!.price && secondToken!.price) {
```

Comment:
these should use optional chaining instead of non-null assertions

```suggestion
        if (adaToken?.price && secondToken?.price) {
```

with the non-null `!` if i.e. `secondToken` is undefined, this will throw an error, with optional chaining `?` it will simply return a falsy value